.github/workflows/rust.yml

```diff
@@ -3,28 +3,58 @@ name: Rust
 on: [push]
 
 jobs:
-  build:
+
+  codestyle:
+    runs-on: ubuntu-latest
+    steps:
+      - name: Set up Rust
+        uses: hecrj/setup-rust-action@v1
+        with:
+          components: rustfmt
+          rust-version: nightly
+      - uses: actions/checkout@v2
+      - run: cargo fmt -- --check --config-path <(echo 'license_template_path = "HEADER"')
+
+  lint:
+    runs-on: ubuntu-latest
+    steps:
+      - name: Set up Rust
+        uses: hecrj/setup-rust-action@v1
+        with:
+          components: clippy
+      - uses: actions/checkout@v2
+      - run: cargo clippy --all-targets -- -D clippy::all
+
+  compile:
     runs-on: ubuntu-latest
+    steps:
+      - name: Set up Rust
+        uses: hecrj/setup-rust-action@v1
+      - uses: actions/checkout@master
+      - run: cargo check --all
 
+  test:
+    needs: [codestyle, lint, compile]
+    strategy:
+      matrix:
+        rust: [stable, beta, nightly]
+    runs-on: ubuntu-latest
     steps:
-    - uses: actions/checkout@v1
     - name: Setup Rust
+      uses: hecrj/setup-rust-action@v1
+      with:
+        rust-version: ${{ matrix.rust }}
+    - name: Checkout
+      uses: actions/checkout@v2
+    - name: Test
+      run: cargo test
+    - name: Coverage
+      if: matrix.rust == 'stable'
       run: |
-        rustup toolchain install nightly --profile default
-        rustup toolchain install stable
-        rustup override set stable
-    # Clippy must be run first, as its lints are only triggered during
-    # compilation. Put another way: after a successful `cargo build`, `cargo
-    # clippy` is guaranteed to produce no results. This bug is known upstream:
-    # https://github.com/rust-lang/rust-clippy/issues/2604.
-#    - name: Clippy
-#      run: cargo clippy -- --all-targets --all-features -- -D warnings
-    - name: Check formatting
-      run: |
-        cargo +nightly fmt -- --check --config-path <(echo 'license_template_path = "HEADER"')
-    - name: Build
-      run: cargo build --verbose
-    - name: Run tests
-      run: cargo test --verbose
-    - name: Run tests for all features
-      run: cargo test --verbose -- all-features
+        # Tarpaulin interoperates with CI services, however Actions is not
+        # currently supported. As a workaround, we can pretend to be Travis.
+        export TRAVIS_JOB_ID=${GITHUB_SHA}
+        export TRAVIS_PULL_REQUEST=false
+        export TRAVIS_BRANCH=${GITHUB_REF##*/}
+        cargo install cargo-tarpaulin
```

Comment:
There are cache facilities in GA but I haven't explored them in much depth. I can certainly take a look at speeding up the builds via caching.

Reply:
That would be appreciated! But I'm happy to merge this as is, if you figure out what's up with coveralls and if we don't yet make it run on PRs, as requested in a nearby thread. The CI speed on PRs is an issue, on the `master` branch - not so much.

Reply:
Thanks--it shouldn't be too difficult to make this step faster. Of course, the bigger question right now is why things aren't making it to Coveralls so I'll try to resolve that first. 🙂

Reply:
The whole workflow now finishes in a minute, good job! Now the only mystery is what's wrong with coveralls..

Reply:
It's very strange because this technique seems to be working for the [`again`](https://github.com/softprops/again/blob/20586d376fff826f5dbbd3a873e8fc1feab6dcb2/.github/workflows/main.yml#L67) crate... I can see [those results](https://coveralls.io/github/softprops/again) in Coveralls.

Reply:
Hm, I'm not sure what to make of [this issue thread](https://github.com/lemurheavy/coveralls-public/issues/281).

Could codecov be an alternative to explore?

Reply:
I tried to set this up on another Rust project but seem to run into the same issue: things appear to be sent to Coveralls successfully but no reports ever appear.

So out of curiosity [I decided to try Codecov](https://github.com/maxcountryman/forma/compare/feature/codecov). [It seems like it works](https://github.com/maxcountryman/forma/runs/747393739). And I can access [the report](https://codecov.io/gh/maxcountryman/forma/tree/d3134dd0b4972a15df12b7ba8f05b10561e22ff6).

Of course if there's a preference for Coveralls I can continue with debugging that.

Reply:
See the other thread, I'm resolving this one as it was about the cache, which now works!